lib/srv/usermgmt.go

```diff
@@ -0,0 +1,261 @@
+	}
+	return &userCloser{
+		username: name,
+		users:    u,
+		backend:  u.backend,
+	}, nil
+}
+
+func (u *HostUserManagment) createGroupIfNotExist(group string) error {
+	_, err := u.backend.LookupGroup(group)
+	if err != nil && err != user.UnknownGroupError(group) {
+		return trace.Wrap(err)
+	}
+	err = u.backend.CreateGroup(group)
+	if trace.IsAlreadyExists(err) {
+		return nil
+	}
+	return trace.Wrap(err)
+}
+
+// DeleteAllUsers deletes all host users in the teleport service group.
+func (u *HostUserManagment) DeleteAllUsers(reg *SessionRegistry) error {
+	users, err := u.backend.GetAllUsers()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	teleportGroup, err := u.backend.LookupGroup(types.TeleportServiceGroup)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var errs []error
+	reg.sessionsMux.Lock()
+	defer reg.sessionsMux.Unlock()
+
+outer:
+	for _, name := range users {
+		// delete all users is only used for periodic deletes and
+		// as a result shouldnt try to delete currently active
+		// sessions as it could cause a race between logging in
+		// and user deletion.
+		for _, sess := range reg.sessions {
+			if sess.login == name {
+				continue outer
+			}
+		}
+		errs = append(errs, u.DeleteUser(name, teleportGroup.Gid))
+	}
+	return trace.NewAggregate(errs...)
+}
+
+// DeleteUser deletes the user only if they are
+// present in the specified group
+func (u *HostUserManagment) DeleteUser(username string, gid string) error {
+	tempUser, err := u.backend.Lookup(username)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	ids, err := u.backend.UserGIDs(tempUser)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	for _, id := range ids {
+		if id == gid {
+			err := u.backend.DeleteUser(username)
+			if errors.Is(err, ErrUserLoggedIn) {
+				log.Debugf("Not deleting user %q, user has another session, or running process", username)
+				return nil
+			}
+			return trace.Wrap(err)
+		}
+	}
+	log.Debugf("User %q not deleted: not a temporary user", username)
+	return nil
+}
+
+// UserCleanup starts a periodic user deletion cleanup loop for
+// users that failed to delete
+func (u *HostUserManagment) UserCleanup(reg *SessionRegistry, s types.Semaphores, lease types.SemaphoreLease) {
+	defer s.CancelSemaphoreLease(u.ctx, lease)
+	if err := u.DeleteAllUsers(reg); err != nil {
+		log.Error("Error during temporary user cleanup: ", err)
+	}
+
+	cleanupTicker := time.NewTicker(time.Minute * 5)
+	defer cleanupTicker.Stop()
+	for {
+		select {
+		case <-cleanupTicker.C:
+			if err := u.DeleteAllUsers(reg); err != nil {
+				log.Error("Error during temporary user cleanup: ", err)
+			}
+		case <-u.ctx.Done():
+			return
+		}
+		lease.Expires = time.Now().Add(time.Minute * 8)
+		if err := s.KeepAliveSemaphoreLease(u.ctx, lease); err != nil {
+			log.Errorf("Error keeping semaphore lease: %s", err)
+			return
+		}
+	}
+}
+
+// Shutdown cancels the UserCleanup loop
+func (u *HostUserManagment) Shutdown() {
+	u.cancel()
+}
```

Comment:
This should probably wait until the UserCleanup loop has actually finished and the semaphore has been released